### DISCUSSION TITLE
Dynamically resizing the float layout before preview toggle

### DISCUSSION THREAD

**Discussion by asmodeus812** [ibhagwan/fzf-lua#1115] [Q&A] [answered]
Hello @ibhagwan, i am looking to resize a floating layout, doubling the height, that would allow me keep the same height for the list of items below the preview, instead of halving it, when preview is on, i have configured the preview to take up 50% of the float size when enabled `veritcal|horizontal = "up:50%"`. 

Looking at the `toggle_preview` source, i can see that it uses `self.winopts` instead of trying to get the current window config / size, when calculating the preview position - `lua/fzf-lua/win.lua:687`, which leads to an issue where the manual resizing done by me before toggle_preview would not be taken into account. 

The same source below works just fine for `split` i just wanted to extend it to be consistent / working for float as well.

```lua
local function toggle_preview_win()
        local preview = fzf_lua.win.__SELF()

        local winid = preview.fzf_winid
        local hidden = preview.preview_hidden
        local height = vim.api.nvim_win_get_height(winid)

        height = hidden and height * 2 or height / 2
        vim.api.nvim_win_set_height(winid, math.ceil(height))

        preview.toggle_preview()
end
```

**Comment by ibhagwan** [comment]
@asmodeus812, could the behavior added in #721 be relevant in this case?
```lua
require("fzf-lua").setup({
  previewers = { builtin = { toggle_behavior = "extend" } },
})
```

**Reply by asmodeus812** [reply]
Could be but is this only applicable to builtin previewers ? I am trying to make this work for both types - external (like bat, or builtin). It works fine when using `split` as i mentioned above. I think taking into account the current height in the code as i mentioned, should do the trick, instead of taking the current height as it does right now.

**Reply by asmodeus812** [reply]
@ibhagwan doing something like that `preview.winopts.height = height` seems to somewhat work, but might be touching too much into the internals which i do not like.

```lua
        local hidden = preview.preview_hidden
        local height = preview.winopts.height

        height = hidden and height * 2 or height / 2

        preview.winopts.height = height
        preview.winopts.row = view_height - height

        preview.toggle_preview()
```

Does not work really well, it toggles open fine, doubles the win height, but when afterwards when toggling off, the height remains the same, and is not reduced / restored.

**Comment by ibhagwan** [comment]
> Looking at the toggle_preview source, i can see that it uses self.winopts instead of trying to get the current window config / size, when calculating the preview position - lua/fzf-lua/win.lua:687, which leads to an issue where the manual resizing done by me before toggle_preview would not be taken into account.

The issue you’re facing is that `toggle_preview`, if hidden, calls `redraw_preview` which then has no awareness of the previous preview height as it’s most likely closed.

> @ibhagwan doing something like that `preview.winopts.height = height` seems to somewhat work, but might be touching too much into the internals which i do not like

You’re correct, this will most likely break and cause all kinds of other unintended behaviors.

I think the best way to address your issue is to use code similar to the preview rotation code, you hijack the `layout` and then call redraw on both main and preview window, as the `toggle_preview_cw` does:
https://github.com/ibhagwan/fzf-lua/blob/b5e08be6835e7ef81d0b90ac58c185e276a771cd/lua/fzf-lua/win.lua#L1272-L1279

**Reply by asmodeus812** [reply]
I think there is an issue with the existing code - after you call toggle_preview, this still returns nil, `self:validate_preview()` neither the `preview_winid` or `border_winid` are set, are those only used for the builtin previewer, if so, we have this `  self.preview_hidden = not self.preview_hidden`, then this line `if self.preview_hidden and self:validate_preview()`,  when the previewer is toggled off, will never enter since validate always returns nil atm for native previewers. And that is what i observe when i toggle it off it does not enter in the `close previewer` branch at all, if i however remove the validate, then it does enter there and the reduced height i have set beforehand correctly reduces the height back to normal.

```lua
function FzfWin.toggle_preview()
  if not _self then return end
  local self = _self
  self.preview_hidden = not self.preview_hidden
  if self.winopts.split and self._fzf_toggle_prev_bind then
    utils.feed_keys_termcodes(self._fzf_toggle_prev_bind)
  end
  if self.preview_hidden then -- removed validate call from here to test out the toggle off, and it does work the height is reduced.
    self:close_preview()
    self:redraw_main()
  elseif not self.preview_hidden then
    self:redraw_main()
    self:redraw_preview()
  end
  -- close_preview() calls FzfWin:close()
  -- which will clear out our singleton so
  -- we must save it again to call redraw
  _self = self
end
```

```lua
if not self.previewer_is_builtin or self.preview_hidden then return end -- well yeah.......
```

**Reply by asmodeus812** [reply]
Feel like that in toggle preview, we should have something like that instead

```lua
            if preview.preview_hidden then
                if preview:validate_preview() then
                    preview:close_preview()
                end
                redraw_main()
            end
```

**Comment by ibhagwan** [comment]
> validate always returns nil atm for native previewers

That is correct because there is no preview window/buffer as the preview is handled by fzf in the terminal.

> Feel like that in toggle preview, we should have something like that instead

I’m not sure what your code replaces but if it doesn’t regress any existing use case and it solves your issue make a PR and I will accept it.

In general the window/previewer code has been through a journey, it started as just a single window manager with only fzf native previewers, then the neovim “builtin” previewer was added on top, then splits and then fzf-tmux.

Due to the above there is a lot of improvement to be had in that part of the code, I would’ve loved for it to be split into classes  per use-case, one for splits, one for native, etc and have a better designed relationship with the previewer class, optimally I’d like to users to be able to supply their window creation class via an API for integrations like nui, etc.

As you can imagine this would take time and effort and as of now no real benefit as the code most of the time works flawlessly so I haven’t gotten to it and don’t know when I will.

**Comment by asmodeus812** [comment]
> I’m not sure what your code replaces but if it doesn’t regress any existing use case and it solves your issue make a PR and I will accept it.

Currently this is the first branch of the toggle_preview, however its wrong

```lua
if preview.preview_hidden and preview:validate_preview() then
        preview:close_preview()
        redraw_main()
end
```

`validate_preview` should be called when preview windows exist, however for native they never do, meaning the branch will fail when toggling native preview off, thus changed slightly to

```lua
   if preview.preview_hidden then
         if preview:validate_preview() then
             preview:close_preview()
         end
      redraw_main()
   end
```

There is an obvious short circuit, in the code that creates the preview & border windows, which pretty much says it all, preview windows are not created for native ones, **which is expected, since those are delegated to fzf**

```lua
if not self.previewer_is_builtin or self.preview_hidden then return end
```

Can we at least adjust toggle_preview as suggested above ?

**Comment by ibhagwan** [comment, marked as answer]
> Can we at least adjust toggle_preview as suggested above ?

Why would we want to redraw main when no change is required to the fzf terminal window? With native previews, the toggle happens internally inside fzf, this function doesn’t even get called for native previewers, nor does the binding for this exist for native previewers:
https://github.com/ibhagwan/fzf-lua/blob/7ecc8e3f04a830356931ad36645a8d6845fb8aff/lua/fzf-lua/win.lua#L44-L51

As you can see from above only help and full screen key maps are set for native previews, therefore the entire discussion about this function is irrelevant (in the context of native previewers).

You already have the win object from `fzf_lua.win.__SELF()` and the toggle function isn’t too many lines, you’re already messing with internals that I mentioned I’d like to change one day so why not create your own function that does exactly what you want and if the API changes in the future we’ll deal with it then.

If you need an internal win function to be exposed let me know and I’ll make it part of the object.

**Comment by ibhagwan** [comment]
> The same source below works just fine for split i just wanted to extend it to be consistent / working for float as well.

If you’re using native previewer this would require a different approach, why not add a postfix to your acrion `{ fn = …, postfix = 'change-preview-window:…' }`  using the fzf action syntax?

You can see how I use these in the recently added `resize` event to auto rotate the preview layout on `flex`:
https://github.com/ibhagwan/fzf-lua/blob/7ecc8e3f04a830356931ad36645a8d6845fb8aff/lua/fzf-lua/core.lua#L356-L366

**Reply by asmodeus812** [reply]
I want to call redraw because i want to change the height of the window. In the end, what i was looking for is the "extend" functionality you linked for builtins to also apply for native too. The root of the problem is that the winopts height includes the preview, its the total height, and you end up with too much wasted space when the preview is off. 

Set the height to something reasonable, like 20% of the current viewpoint, and when the preview is toggled on both the preview and the list are roo small to be useful. Set the height to 50% of the viewpoint, now the list is too big just to accommodate for an occasional preview toggle. 

Say i want a height of 10 for the fzf list, to always be 10, and the preview should render on top of this. As of now, there is no way to do this without manually modifying the size of the window, as i have shown above. That is why i want to call render main again  through the toggle preview. It is kind of a mess tbh. 

Never mind iwill adress this in my config. It's too much effort spent discussing this, it is not going anywhere. Closing the discussion as resolved.
